File: examples/csharp/Pmg/HandFlow/HandFlowPM.cs
using Angzarr.Client;
using Angzarr.Proto.Angzarr;
using Angzarr.Proto.Examples;

namespace Angzarr.Examples.PmgHandFlow;

/// <summary>
/// Hand Flow Process Manager - orchestrates poker hand phases across domains.
///
/// This PM coordinates the workflow between table and hand domains,
/// tracking phase transitions and dispatching commands as the hand progresses.
/// </summary>

// docs:start:pm_state
public enum HandPhase { AwaitingDeal, Dealing, Blinds, Betting, Complete }

public class HandFlowState
{
    public string HandId { get; set; } = "";
    public HandPhase Phase { get; set; } = HandPhase.AwaitingDeal;
    public int PlayerCount { get; set; } = 0;
}
// docs:end:pm_state

// docs:start:pm_handler
public class HandFlowPM : ProcessManager<HandFlowState>
{
    [ReactsTo(typeof(HandStarted))]
    public IEnumerable<CommandBook> HandleHandStarted(HandStarted evt, HandFlowState state)
    {
        state.HandId = evt.HandId;
        state.Phase = HandPhase.Dealing;
        state.PlayerCount = evt.PlayerCount;

        yield return BuildCommand("hand", new DealCards
        {
            HandId = evt.HandId,
            PlayerCount = evt.PlayerCount
        });
    }

    [ReactsTo(typeof(CardsDealt))]
    public IEnumerable<CommandBook> HandleCardsDealt(CardsDealt evt, HandFlowState state)
    {
        state.Phase = HandPhase.Blinds;
        yield return BuildCommand("hand", new PostBlinds { HandId = state.HandId });
    }

    [ReactsTo(typeof(HandComplete))]
    public IEnumerable<CommandBook> HandleHandComplete(HandComplete evt, HandFlowState state)
    {
        state.Phase = HandPhase.Complete;
        yield return BuildCommand("table", new EndHand
        {
            HandId = state.HandId,
            WinnerId = evt.WinnerId
        });
    }
}
// docs:end:pm_handler

File: examples/java/pmg-hand-flow/src/main/java/dev/angzarr/examples/pmghandflow/HandFlowPM.java
package dev.angzarr.examples.pmghandflow;

import dev.angzarr.client.ProcessManager;
import dev.angzarr.proto.angzarr.CommandBook;
import dev.angzarr.proto.examples.CardsDealt;
import dev.angzarr.proto.examples.DealCards;
import dev.angzarr.proto.examples.EndHand;
import dev.angzarr.proto.examples.HandComplete;
import dev.angzarr.proto.examples.HandStarted;
import dev.angzarr.proto.examples.PostBlinds;

import java.util.List;

/**
 * Hand Flow Process Manager - orchestrates poker hand phases across domains.
 *
 * This PM coordinates the workflow between table and hand domains,
 * tracking phase transitions and dispatching commands as the hand progresses.
 */

// docs:start:pm_state
enum HandPhase {
    AWAITING_DEAL, DEALING, BLINDS, BETTING, COMPLETE
}

class HandFlowState {
    private String handId = "";
    private HandPhase phase = HandPhase.AWAITING_DEAL;
    private int playerCount = 0;

    public String getHandId() { return handId; }
    public void setHandId(String handId) { this.handId = handId; }
    public HandPhase getPhase() { return phase; }
    public void setPhase(HandPhase phase) { this.phase = phase; }
    public int getPlayerCount() { return playerCount; }
    public void setPlayerCount(int playerCount) { this.playerCount = playerCount; }
}
// docs:end:pm_state

// docs:start:pm_handler
public class HandFlowPM extends ProcessManager<HandFlowState> {

    @ReactsTo(HandStarted.class)
    public List<CommandBook> handleHandStarted(HandStarted event, HandFlowState state) {
        state.setHandId(event.getHandId());
        state.setPhase(HandPhase.DEALING);
        state.setPlayerCount(event.getPlayerCount());

        return List.of(buildCommand("hand", DealCards.newBuilder()
            .setHandId(event.getHandId())
            .setPlayerCount(event.getPlayerCount())
            .build()));
    }

    @ReactsTo(CardsDealt.class)
    public List<CommandBook> handleCardsDealt(CardsDealt event, HandFlowState state) {
        state.setPhase(HandPhase.BLINDS);
        return List.of(buildCommand("hand", PostBlinds.newBuilder()
            .setHandId(state.getHandId())
            .build()));
    }

    @ReactsTo(HandComplete.class)
    public List<CommandBook> handleHandComplete(HandComplete event, HandFlowState state) {
        state.setPhase(HandPhase.COMPLETE);
        return List.of(buildCommand("table", EndHand.newBuilder()
            .setHandId(state.getHandId())
            .setWinnerId(event.getWinnerId())
            .build()));
    }
}
// docs:end:pm_handler
